In [1]:
import numpy as np
import pandas as pd
import csv 

In [2]:
DF = pd.read_csv("IPL_Twitter_MissingData.csv")#,header=None)

In [3]:
print(DF.head())
print(type(DF))

    Q1   Q2      X1       X2       X3       X4
0  0.0  0.0  8130.0   4838.0   3623.0  12718.0
1  0.0  0.0  2594.0   3651.0   2323.0   6014.0
2  1.0  1.0  4108.0  12510.0  12127.0   3816.0
3  1.0  0.0  2171.0      NaN   6732.0   1357.0
4  0.0  0.0  4518.0   4326.0   4214.0   6862.0
<class 'pandas.core.frame.DataFrame'>


In [4]:
df = pd.DataFrame({'a':[1,2,np.nan], 'b':[np.nan,1,np.nan], 'c' : [1,1,1]})
print(df)
print(df.isna().sum())

     a    b  c
0  1.0  NaN  1
1  2.0  1.0  1
2  NaN  NaN  1
a    1
b    2
c    0
dtype: int64


In [5]:
DF.isna().sum(axis = 1)
row_wise_nanvalues = pd.DataFrame(DF.isna().sum(axis = 1))
print(row_wise_nanvalues[0] > 0)
row_missing_value = pd.DataFrame(row_wise_nanvalues[0] > 0) #like boolean in np boolean masking)
print(row_missing_value)
DF1 =DF.copy()
DF2 = DF.copy()
print(type(row_missing_value))
# DF1['row has nan?'] = row_missing_value[0]
# print(DF1.head())
# print(DF.head())

0      False
1      False
2      False
3       True
4      False
5      False
6       True
7      False
8      False
9       True
10     False
11     False
12     False
13     False
14     False
15     False
16     False
17     False
18     False
19     False
20     False
21      True
22     False
23     False
24     False
25      True
26      True
27      True
28     False
29     False
       ...  
970    False
971    False
972     True
973    False
974    False
975    False
976     True
977    False
978     True
979    False
980    False
981    False
982    False
983    False
984     True
985    False
986     True
987    False
988    False
989     True
990    False
991    False
992    False
993    False
994     True
995    False
996    False
997    False
998    False
999    False
Name: 0, Length: 1000, dtype: bool
         0
0    False
1    False
2    False
3     True
4    False
5    False
6     True
7    False
8    False
9     True
10   False
11   False
12   False
13   False
14   Fa

In [6]:
series = pd.Series([True,False,True,False,False,False,True])
series
print(series.values)
type(series.values)

[ True False  True False False False  True]


numpy.ndarray

In [7]:
nan_samples = np.sum(row_missing_value[0])
nan_samples
correct_samples = (DF2.shape)[0]-nan_samples
# print(type(nan_samples))
print('Number of samples with missing values are ',nan_samples, '.', sep ='' )
print('Number of samples with out missing values are ',correct_samples,'.', sep ='' )
print('As total number of samples are ',(DF2.shape)[0],'.', sep='')

Number of samples with missing values are 238.
Number of samples with out missing values are 762.
As total number of samples are 1000.


In [8]:
column_wise_nanvalues = pd.DataFrame(DF.isnull().sum(axis = 0))
type(column_wise_nanvalues) #data frame
column_wise_nanvalues.columns = ['number of values missing from each column']
column_wise_nanvalues

number of values missing from each column
Q1                                         44
Q2                                         45
X1                                         36
X2                                         55
X3                                         67
X4                                         41

## 1) How many data samples have missing values? 

if the column vector (Q1, Q2, X1, X2, X3, X4) represents a sample, then having at least one missing value in the sample includes it in missing values sample.

so, it is 238 samples with missing values and 1000-238 = 762 samples have complete data

In [9]:
# Df_without_nan = DF.where(DF['row has nan?'] == False)
# print(Df_without_nan.head())
# DF.shape

In [10]:
DF = DF.dropna()

In [11]:
print(DF.shape)
# DF.dropna().shape

(762, 6)


## 4. Impute the missing values using your own ideas and also briefly explain those ideas in 
the report (scores will be proportional to the number of meaningful ways the data is 
filled). One has to use one unique idea to fill the complete data so submit multiple sets 
of filled data, if you have multiple ideas for filling the dat## 4.

### filling it with a constant for Q1 and Q2 as they have to be binary

In [12]:
hhuh = pd.DataFrame([0,0,0,1,1,1,0,1,1,0,0,1,1,np.nan,np.nan,0,1,0,np.nan,1,1,0,1,0,1,np.nan,1,1])
# hhuh.interpolate(method ='linear', limit_direction ='backward', limit = 1) 
# df.interpolate?

In [13]:
# huh = pd.DataFrame([16,np.nan,36,0,1,4,9,49,64,np.nan,np.nan,121,144])
# huh.interpolate(method ='quadratic', limit_direction ='backward', limit = 1)  this won't work because the function is approximated as quadratic


In [14]:
#fill Q1 and Q2 by 1 if the number of 1 are more than 0's that is if (sum of entries in Q1/total entries after deleting nan entries in the column)
temp_col = DF2['Q1'].dropna()
checkQ1 = sum(temp_col)/temp_col.size
checkQ2 = sum(DF2['Q2'].dropna())/(DF2['Q1'].dropna().size)
print(checkQ1,checkQ2)
if checkQ1>=0.5:
    DF2['Q1'].fillna(1)
else:
    DF2['Q1'] = DF2['Q1'].fillna(0)
#     pass
if checkQ2>=0.5:
    DF2['Q2'].fillna(1)
else:
    DF2['Q2'] = DF2['Q2'].fillna(0)
#     pass
DF2

0.4989539748953975 0.49476987447698745


Q1   Q2      X1       X2       X3       X4
0    0.0  0.0  8130.0   4838.0   3623.0  12718.0
1    0.0  0.0  2594.0   3651.0   2323.0   6014.0
2    1.0  1.0  4108.0  12510.0  12127.0   3816.0
3    1.0  0.0  2171.0      NaN   6732.0   1357.0
4    0.0  0.0  4518.0   4326.0   4214.0   6862.0
5    0.0  1.0  4218.0   9931.0  10149.0   5324.0
6    0.0  0.0  1916.0   8210.0  12422.0   3383.0
7    0.0  1.0  1470.0   8204.0  11644.0   3761.0
8    0.0  0.0  7109.0   4031.0   3973.0   9715.0
9    1.0  0.0  2728.0      NaN   8562.0   5144.0
10   0.0  1.0  4584.0   5495.0  10220.0    772.0
11   1.0  1.0  3031.0  14284.0  11921.0   4202.0
12   1.0  0.0  4366.0  14286.0   9819.0   5476.0
13   1.0  0.0  4572.0  11782.0   7732.0   3390.0
14   1.0  0.0  2293.0  13848.0   6170.0    419.0
15   1.0  0.0  2585.0  13526.0   7793.0   3328.0
16   1.0  1.0  2772.0  12740.0  12834.0   2340.0
17   1.0  0.0  2284.0  14148.0   9111.0   5370.0
18   0.0  1.0  2969.0   8474.0  14864.0   2529.0
19   0.0  0.0  9818.0   2755.0   1444.0  14121.0
20   0.0  1.0  3960.0   6174.0  13675.0    398.0
21   0.0  1.0  3451.0   9097.0  12659.0   3804.0
22   1.0  0.0  3278.0  10307.0   7481.0   4581.0
23   1.0  1.0  1381.0  11763.0  12967.0    394.0
24   0.0  1.0  4952.0   9321.0  11944.0   3903.0
25   1.0  0.0  2805.0      NaN      NaN   6316.0
26   0.0  0.0  7796.0   2509.0      NaN  10792.0
27   1.0  0.0  1690.0      NaN   6276.0   3171.0
28   1.0  1.0  2870.0  10706.0  10340.0   2731.0
29   0.0  1.0  3209.0   6090.0  13862.0    436.0
..   ...  ...     ...      ...      ...      ...
970  1.0  1.0  2895.0  14756.0  11245.0   4850.0
971  0.0  0.0  5620.0   1710.0   2594.0   5948.0
972  1.0  0.0     NaN      NaN   7901.0   5543.0
973  1.0  1.0  3560.0  10660.0  12264.0   2016.0
974  0.0  0.0  1888.0   2047.0   2341.0   2493.0
975  0.0  1.0  3786.0   5626.0  10650.0    952.0
976  1.0  1.0  4511.0      NaN  12187.0   4015.0
977  1.0  1.0  2001.0  14057.0  10422.0   4158.0
978  0.0  0.0  7890.0   1754.0   2150.0   9262.0
979  0.0  1.0  4989.0   6121.0  13262.0    232.0
980  0.0  1.0  1897.0   6345.0  13365.0   1193.0
981  0.0  1.0  3881.0   6734.0  12585.0   1359.0
982  0.0  0.0  8657.0   3242.0   4719.0   9211.0
983  0.0  1.0  2805.0   6204.0  13575.0    752.0
984  1.0  0.0  2608.0      NaN   6532.0   2476.0
985  0.0  0.0  7085.0   2156.0   3687.0   6943.0
986  1.0  0.0  3797.0  13982.0   7208.0      NaN
987  0.0  1.0  1449.0   8922.0  11458.0   4549.0
988  0.0  0.0  9087.0   1472.0   4954.0   6449.0
989  0.0  0.0  4755.0      NaN      NaN   1396.0
990  1.0  1.0  3935.0  12186.0  11899.0   3602.0
991  1.0  1.0  3341.0  13339.0  13240.0   2962.0
992  0.0  1.0  4656.0   8534.0  12789.0   2894.0
993  0.0  0.0  4352.0   1792.0   2959.0   4119.0
994  1.0  1.0  3048.0      NaN      NaN   3771.0
995  1.0  0.0  2161.0  13773.0   7791.0   3372.0
996  0.0  1.0  3779.0   7132.0  14182.0   1224.0
997  0.0  0.0  5919.0   2703.0   3578.0   6640.0
998  1.0  1.0  3825.0  14157.0  10174.0   5970.0
999  1.0  1.0  3847.0  14364.0  14690.0   3117.0

[1000 rows x 6 columns]

In [15]:
DF3 = pd.read_csv("IPL_Twitter_MissingData.csv")
DF4 = DF3.copy()
DF3

Q1   Q2      X1       X2       X3       X4
0    0.0  0.0  8130.0   4838.0   3623.0  12718.0
1    0.0  0.0  2594.0   3651.0   2323.0   6014.0
2    1.0  1.0  4108.0  12510.0  12127.0   3816.0
3    1.0  0.0  2171.0      NaN   6732.0   1357.0
4    0.0  0.0  4518.0   4326.0   4214.0   6862.0
5    0.0  1.0  4218.0   9931.0  10149.0   5324.0
6    0.0  NaN  1916.0   8210.0  12422.0   3383.0
7    0.0  1.0  1470.0   8204.0  11644.0   3761.0
8    0.0  0.0  7109.0   4031.0   3973.0   9715.0
9    1.0  0.0  2728.0      NaN   8562.0   5144.0
10   0.0  1.0  4584.0   5495.0  10220.0    772.0
11   1.0  1.0  3031.0  14284.0  11921.0   4202.0
12   1.0  0.0  4366.0  14286.0   9819.0   5476.0
13   1.0  0.0  4572.0  11782.0   7732.0   3390.0
14   1.0  0.0  2293.0  13848.0   6170.0    419.0
15   1.0  0.0  2585.0  13526.0   7793.0   3328.0
16   1.0  1.0  2772.0  12740.0  12834.0   2340.0
17   1.0  0.0  2284.0  14148.0   9111.0   5370.0
18   0.0  1.0  2969.0   8474.0  14864.0   2529.0
19   0.0  0.0  9818.0   2755.0   1444.0  14121.0
20   0.0  1.0  3960.0   6174.0  13675.0    398.0
21   NaN  1.0  3451.0   9097.0  12659.0   3804.0
22   1.0  0.0  3278.0  10307.0   7481.0   4581.0
23   1.0  1.0  1381.0  11763.0  12967.0    394.0
24   0.0  1.0  4952.0   9321.0  11944.0   3903.0
25   1.0  0.0  2805.0      NaN      NaN   6316.0
26   0.0  0.0  7796.0   2509.0      NaN  10792.0
27   1.0  0.0  1690.0      NaN   6276.0   3171.0
28   1.0  1.0  2870.0  10706.0  10340.0   2731.0
29   0.0  1.0  3209.0   6090.0  13862.0    436.0
..   ...  ...     ...      ...      ...      ...
970  1.0  1.0  2895.0  14756.0  11245.0   4850.0
971  0.0  0.0  5620.0   1710.0   2594.0   5948.0
972  1.0  0.0     NaN      NaN   7901.0   5543.0
973  1.0  1.0  3560.0  10660.0  12264.0   2016.0
974  0.0  0.0  1888.0   2047.0   2341.0   2493.0
975  0.0  1.0  3786.0   5626.0  10650.0    952.0
976  1.0  1.0  4511.0      NaN  12187.0   4015.0
977  1.0  1.0  2001.0  14057.0  10422.0   4158.0
978  0.0  NaN  7890.0   1754.0   2150.0   9262.0
979  0.0  1.0  4989.0   6121.0  13262.0    232.0
980  0.0  1.0  1897.0   6345.0  13365.0   1193.0
981  0.0  1.0  3881.0   6734.0  12585.0   1359.0
982  0.0  0.0  8657.0   3242.0   4719.0   9211.0
983  0.0  1.0  2805.0   6204.0  13575.0    752.0
984  1.0  0.0  2608.0      NaN   6532.0   2476.0
985  0.0  0.0  7085.0   2156.0   3687.0   6943.0
986  1.0  0.0  3797.0  13982.0   7208.0      NaN
987  0.0  1.0  1449.0   8922.0  11458.0   4549.0
988  0.0  0.0  9087.0   1472.0   4954.0   6449.0
989  0.0  0.0  4755.0      NaN      NaN   1396.0
990  1.0  1.0  3935.0  12186.0  11899.0   3602.0
991  1.0  1.0  3341.0  13339.0  13240.0   2962.0
992  0.0  1.0  4656.0   8534.0  12789.0   2894.0
993  0.0  0.0  4352.0   1792.0   2959.0   4119.0
994  1.0  1.0  3048.0      NaN      NaN   3771.0
995  1.0  0.0  2161.0  13773.0   7791.0   3372.0
996  0.0  1.0  3779.0   7132.0  14182.0   1224.0
997  0.0  0.0  5919.0   2703.0   3578.0   6640.0
998  1.0  1.0  3825.0  14157.0  10174.0   5970.0
999  1.0  1.0  3847.0  14364.0  14690.0   3117.0

[1000 rows x 6 columns]

### ffill and bfill

In [16]:
DF4[['Q1','Q2']] = DF4[['Q1','Q2']].bfill()
DF4 #Q1 column 21 is filled with 0 (20 is 0, 21 is 1) and Q2 column 6 is filled with 1 (5,7 both are 1s)

Q1   Q2      X1       X2       X3       X4
0    0.0  0.0  8130.0   4838.0   3623.0  12718.0
1    0.0  0.0  2594.0   3651.0   2323.0   6014.0
2    1.0  1.0  4108.0  12510.0  12127.0   3816.0
3    1.0  0.0  2171.0      NaN   6732.0   1357.0
4    0.0  0.0  4518.0   4326.0   4214.0   6862.0
5    0.0  1.0  4218.0   9931.0  10149.0   5324.0
6    0.0  1.0  1916.0   8210.0  12422.0   3383.0
7    0.0  1.0  1470.0   8204.0  11644.0   3761.0
8    0.0  0.0  7109.0   4031.0   3973.0   9715.0
9    1.0  0.0  2728.0      NaN   8562.0   5144.0
10   0.0  1.0  4584.0   5495.0  10220.0    772.0
11   1.0  1.0  3031.0  14284.0  11921.0   4202.0
12   1.0  0.0  4366.0  14286.0   9819.0   5476.0
13   1.0  0.0  4572.0  11782.0   7732.0   3390.0
14   1.0  0.0  2293.0  13848.0   6170.0    419.0
15   1.0  0.0  2585.0  13526.0   7793.0   3328.0
16   1.0  1.0  2772.0  12740.0  12834.0   2340.0
17   1.0  0.0  2284.0  14148.0   9111.0   5370.0
18   0.0  1.0  2969.0   8474.0  14864.0   2529.0
19   0.0  0.0  9818.0   2755.0   1444.0  14121.0
20   0.0  1.0  3960.0   6174.0  13675.0    398.0
21   0.0  1.0  3451.0   9097.0  12659.0   3804.0
22   1.0  0.0  3278.0  10307.0   7481.0   4581.0
23   1.0  1.0  1381.0  11763.0  12967.0    394.0
24   0.0  1.0  4952.0   9321.0  11944.0   3903.0
25   1.0  0.0  2805.0      NaN      NaN   6316.0
26   0.0  0.0  7796.0   2509.0      NaN  10792.0
27   1.0  0.0  1690.0      NaN   6276.0   3171.0
28   1.0  1.0  2870.0  10706.0  10340.0   2731.0
29   0.0  1.0  3209.0   6090.0  13862.0    436.0
..   ...  ...     ...      ...      ...      ...
970  1.0  1.0  2895.0  14756.0  11245.0   4850.0
971  0.0  0.0  5620.0   1710.0   2594.0   5948.0
972  1.0  0.0     NaN      NaN   7901.0   5543.0
973  1.0  1.0  3560.0  10660.0  12264.0   2016.0
974  0.0  0.0  1888.0   2047.0   2341.0   2493.0
975  0.0  1.0  3786.0   5626.0  10650.0    952.0
976  1.0  1.0  4511.0      NaN  12187.0   4015.0
977  1.0  1.0  2001.0  14057.0  10422.0   4158.0
978  0.0  1.0  7890.0   1754.0   2150.0   9262.0
979  0.0  1.0  4989.0   6121.0  13262.0    232.0
980  0.0  1.0  1897.0   6345.0  13365.0   1193.0
981  0.0  1.0  3881.0   6734.0  12585.0   1359.0
982  0.0  0.0  8657.0   3242.0   4719.0   9211.0
983  0.0  1.0  2805.0   6204.0  13575.0    752.0
984  1.0  0.0  2608.0      NaN   6532.0   2476.0
985  0.0  0.0  7085.0   2156.0   3687.0   6943.0
986  1.0  0.0  3797.0  13982.0   7208.0      NaN
987  0.0  1.0  1449.0   8922.0  11458.0   4549.0
988  0.0  0.0  9087.0   1472.0   4954.0   6449.0
989  0.0  0.0  4755.0      NaN      NaN   1396.0
990  1.0  1.0  3935.0  12186.0  11899.0   3602.0
991  1.0  1.0  3341.0  13339.0  13240.0   2962.0
992  0.0  1.0  4656.0   8534.0  12789.0   2894.0
993  0.0  0.0  4352.0   1792.0   2959.0   4119.0
994  1.0  1.0  3048.0      NaN      NaN   3771.0
995  1.0  0.0  2161.0  13773.0   7791.0   3372.0
996  0.0  1.0  3779.0   7132.0  14182.0   1224.0
997  0.0  0.0  5919.0   2703.0   3578.0   6640.0
998  1.0  1.0  3825.0  14157.0  10174.0   5970.0
999  1.0  1.0  3847.0  14364.0  14690.0   3117.0

[1000 rows x 6 columns]

In [17]:
DF3[['Q1','Q2']] = DF3[['Q1','Q2']].ffill() # here Q1 21 is filled by 1 as this is backward filling

In [18]:
DF4

Q1   Q2      X1       X2       X3       X4
0    0.0  0.0  8130.0   4838.0   3623.0  12718.0
1    0.0  0.0  2594.0   3651.0   2323.0   6014.0
2    1.0  1.0  4108.0  12510.0  12127.0   3816.0
3    1.0  0.0  2171.0      NaN   6732.0   1357.0
4    0.0  0.0  4518.0   4326.0   4214.0   6862.0
5    0.0  1.0  4218.0   9931.0  10149.0   5324.0
6    0.0  1.0  1916.0   8210.0  12422.0   3383.0
7    0.0  1.0  1470.0   8204.0  11644.0   3761.0
8    0.0  0.0  7109.0   4031.0   3973.0   9715.0
9    1.0  0.0  2728.0      NaN   8562.0   5144.0
10   0.0  1.0  4584.0   5495.0  10220.0    772.0
11   1.0  1.0  3031.0  14284.0  11921.0   4202.0
12   1.0  0.0  4366.0  14286.0   9819.0   5476.0
13   1.0  0.0  4572.0  11782.0   7732.0   3390.0
14   1.0  0.0  2293.0  13848.0   6170.0    419.0
15   1.0  0.0  2585.0  13526.0   7793.0   3328.0
16   1.0  1.0  2772.0  12740.0  12834.0   2340.0
17   1.0  0.0  2284.0  14148.0   9111.0   5370.0
18   0.0  1.0  2969.0   8474.0  14864.0   2529.0
19   0.0  0.0  9818.0   2755.0   1444.0  14121.0
20   0.0  1.0  3960.0   6174.0  13675.0    398.0
21   0.0  1.0  3451.0   9097.0  12659.0   3804.0
22   1.0  0.0  3278.0  10307.0   7481.0   4581.0
23   1.0  1.0  1381.0  11763.0  12967.0    394.0
24   0.0  1.0  4952.0   9321.0  11944.0   3903.0
25   1.0  0.0  2805.0      NaN      NaN   6316.0
26   0.0  0.0  7796.0   2509.0      NaN  10792.0
27   1.0  0.0  1690.0      NaN   6276.0   3171.0
28   1.0  1.0  2870.0  10706.0  10340.0   2731.0
29   0.0  1.0  3209.0   6090.0  13862.0    436.0
..   ...  ...     ...      ...      ...      ...
970  1.0  1.0  2895.0  14756.0  11245.0   4850.0
971  0.0  0.0  5620.0   1710.0   2594.0   5948.0
972  1.0  0.0     NaN      NaN   7901.0   5543.0
973  1.0  1.0  3560.0  10660.0  12264.0   2016.0
974  0.0  0.0  1888.0   2047.0   2341.0   2493.0
975  0.0  1.0  3786.0   5626.0  10650.0    952.0
976  1.0  1.0  4511.0      NaN  12187.0   4015.0
977  1.0  1.0  2001.0  14057.0  10422.0   4158.0
978  0.0  1.0  7890.0   1754.0   2150.0   9262.0
979  0.0  1.0  4989.0   6121.0  13262.0    232.0
980  0.0  1.0  1897.0   6345.0  13365.0   1193.0
981  0.0  1.0  3881.0   6734.0  12585.0   1359.0
982  0.0  0.0  8657.0   3242.0   4719.0   9211.0
983  0.0  1.0  2805.0   6204.0  13575.0    752.0
984  1.0  0.0  2608.0      NaN   6532.0   2476.0
985  0.0  0.0  7085.0   2156.0   3687.0   6943.0
986  1.0  0.0  3797.0  13982.0   7208.0      NaN
987  0.0  1.0  1449.0   8922.0  11458.0   4549.0
988  0.0  0.0  9087.0   1472.0   4954.0   6449.0
989  0.0  0.0  4755.0      NaN      NaN   1396.0
990  1.0  1.0  3935.0  12186.0  11899.0   3602.0
991  1.0  1.0  3341.0  13339.0  13240.0   2962.0
992  0.0  1.0  4656.0   8534.0  12789.0   2894.0
993  0.0  0.0  4352.0   1792.0   2959.0   4119.0
994  1.0  1.0  3048.0      NaN      NaN   3771.0
995  1.0  0.0  2161.0  13773.0   7791.0   3372.0
996  0.0  1.0  3779.0   7132.0  14182.0   1224.0
997  0.0  0.0  5919.0   2703.0   3578.0   6640.0
998  1.0  1.0  3825.0  14157.0  10174.0   5970.0
999  1.0  1.0  3847.0  14364.0  14690.0   3117.0

[1000 rows x 6 columns]

### Filling other columns (integers)

fill by constant. By median or mean

In [19]:
# print(DF4['X1'].mean())
DF5 = DF2.copy()
DF6 = DF2.copy()
DF6 = DF6.interpolate(method='linear')
DF2['X1'] = DF2['X1'].fillna(int(DF2['X1'].mean()))
DF2['X2'] = DF2['X2'].fillna(int(DF2['X2'].mean()))
DF2['X3'] = DF2['X3'].fillna(int(DF2['X3'].mean()))
DF2['X4'] = DF2['X4'].fillna(int(DF2['X4'].mean()))
DF5['X1'] = DF5['X1'].fillna(int(DF5['X1'].median()))
DF5['X2'] = DF5['X2'].fillna(int(DF5['X2'].median()))
DF5['X3'] = DF5['X3'].fillna(int(DF5['X3'].median()))
DF5['X4'] = DF5['X4'].fillna(int(DF5['X4'].median()))

In [20]:
DF5

Q1   Q2      X1       X2       X3       X4
0    0.0  0.0  8130.0   4838.0   3623.0  12718.0
1    0.0  0.0  2594.0   3651.0   2323.0   6014.0
2    1.0  1.0  4108.0  12510.0  12127.0   3816.0
3    1.0  0.0  2171.0   9924.0   6732.0   1357.0
4    0.0  0.0  4518.0   4326.0   4214.0   6862.0
5    0.0  1.0  4218.0   9931.0  10149.0   5324.0
6    0.0  0.0  1916.0   8210.0  12422.0   3383.0
7    0.0  1.0  1470.0   8204.0  11644.0   3761.0
8    0.0  0.0  7109.0   4031.0   3973.0   9715.0
9    1.0  0.0  2728.0   9924.0   8562.0   5144.0
10   0.0  1.0  4584.0   5495.0  10220.0    772.0
11   1.0  1.0  3031.0  14284.0  11921.0   4202.0
12   1.0  0.0  4366.0  14286.0   9819.0   5476.0
13   1.0  0.0  4572.0  11782.0   7732.0   3390.0
14   1.0  0.0  2293.0  13848.0   6170.0    419.0
15   1.0  0.0  2585.0  13526.0   7793.0   3328.0
16   1.0  1.0  2772.0  12740.0  12834.0   2340.0
17   1.0  0.0  2284.0  14148.0   9111.0   5370.0
18   0.0  1.0  2969.0   8474.0  14864.0   2529.0
19   0.0  0.0  9818.0   2755.0   1444.0  14121.0
20   0.0  1.0  3960.0   6174.0  13675.0    398.0
21   0.0  1.0  3451.0   9097.0  12659.0   3804.0
22   1.0  0.0  3278.0  10307.0   7481.0   4581.0
23   1.0  1.0  1381.0  11763.0  12967.0    394.0
24   0.0  1.0  4952.0   9321.0  11944.0   3903.0
25   1.0  0.0  2805.0   9924.0  10036.0   6316.0
26   0.0  0.0  7796.0   2509.0  10036.0  10792.0
27   1.0  0.0  1690.0   9924.0   6276.0   3171.0
28   1.0  1.0  2870.0  10706.0  10340.0   2731.0
29   0.0  1.0  3209.0   6090.0  13862.0    436.0
..   ...  ...     ...      ...      ...      ...
970  1.0  1.0  2895.0  14756.0  11245.0   4850.0
971  0.0  0.0  5620.0   1710.0   2594.0   5948.0
972  1.0  0.0  3399.0   9924.0   7901.0   5543.0
973  1.0  1.0  3560.0  10660.0  12264.0   2016.0
974  0.0  0.0  1888.0   2047.0   2341.0   2493.0
975  0.0  1.0  3786.0   5626.0  10650.0    952.0
976  1.0  1.0  4511.0   9924.0  12187.0   4015.0
977  1.0  1.0  2001.0  14057.0  10422.0   4158.0
978  0.0  0.0  7890.0   1754.0   2150.0   9262.0
979  0.0  1.0  4989.0   6121.0  13262.0    232.0
980  0.0  1.0  1897.0   6345.0  13365.0   1193.0
981  0.0  1.0  3881.0   6734.0  12585.0   1359.0
982  0.0  0.0  8657.0   3242.0   4719.0   9211.0
983  0.0  1.0  2805.0   6204.0  13575.0    752.0
984  1.0  0.0  2608.0   9924.0   6532.0   2476.0
985  0.0  0.0  7085.0   2156.0   3687.0   6943.0
986  1.0  0.0  3797.0  13982.0   7208.0   3477.0
987  0.0  1.0  1449.0   8922.0  11458.0   4549.0
988  0.0  0.0  9087.0   1472.0   4954.0   6449.0
989  0.0  0.0  4755.0   9924.0  10036.0   1396.0
990  1.0  1.0  3935.0  12186.0  11899.0   3602.0
991  1.0  1.0  3341.0  13339.0  13240.0   2962.0
992  0.0  1.0  4656.0   8534.0  12789.0   2894.0
993  0.0  0.0  4352.0   1792.0   2959.0   4119.0
994  1.0  1.0  3048.0   9924.0  10036.0   3771.0
995  1.0  0.0  2161.0  13773.0   7791.0   3372.0
996  0.0  1.0  3779.0   7132.0  14182.0   1224.0
997  0.0  0.0  5919.0   2703.0   3578.0   6640.0
998  1.0  1.0  3825.0  14157.0  10174.0   5970.0
999  1.0  1.0  3847.0  14364.0  14690.0   3117.0

[1000 rows x 6 columns]

In [21]:
# tempdf = pd.DataFrame([np.random.randint(1,100)+np.random.random(1) for i in range(10)])
# tempdf.round(1)

In [22]:
DF3[['X1','X2','X3','X4']] = DF3[['X1','X2','X3','X4']].ffill()
DF4[['X1','X2','X3','X4']] = DF4[['X1','X2','X3','X4']].bfill()

In [23]:
DF6

Q1   Q2      X1       X2            X3       X4
0    0.0  0.0  8130.0   4838.0   3623.000000  12718.0
1    0.0  0.0  2594.0   3651.0   2323.000000   6014.0
2    1.0  1.0  4108.0  12510.0  12127.000000   3816.0
3    1.0  0.0  2171.0   8418.0   6732.000000   1357.0
4    0.0  0.0  4518.0   4326.0   4214.000000   6862.0
5    0.0  1.0  4218.0   9931.0  10149.000000   5324.0
6    0.0  0.0  1916.0   8210.0  12422.000000   3383.0
7    0.0  1.0  1470.0   8204.0  11644.000000   3761.0
8    0.0  0.0  7109.0   4031.0   3973.000000   9715.0
9    1.0  0.0  2728.0   4763.0   8562.000000   5144.0
10   0.0  1.0  4584.0   5495.0  10220.000000    772.0
11   1.0  1.0  3031.0  14284.0  11921.000000   4202.0
12   1.0  0.0  4366.0  14286.0   9819.000000   5476.0
13   1.0  0.0  4572.0  11782.0   7732.000000   3390.0
14   1.0  0.0  2293.0  13848.0   6170.000000    419.0
15   1.0  0.0  2585.0  13526.0   7793.000000   3328.0
16   1.0  1.0  2772.0  12740.0  12834.000000   2340.0
17   1.0  0.0  2284.0  14148.0   9111.000000   5370.0
18   0.0  1.0  2969.0   8474.0  14864.000000   2529.0
19   0.0  0.0  9818.0   2755.0   1444.000000  14121.0
20   0.0  1.0  3960.0   6174.0  13675.000000    398.0
21   0.0  1.0  3451.0   9097.0  12659.000000   3804.0
22   1.0  0.0  3278.0  10307.0   7481.000000   4581.0
23   1.0  1.0  1381.0  11763.0  12967.000000    394.0
24   0.0  1.0  4952.0   9321.0  11944.000000   3903.0
25   1.0  0.0  2805.0   5915.0  10054.666667   6316.0
26   0.0  0.0  7796.0   2509.0   8165.333333  10792.0
27   1.0  0.0  1690.0   6607.5   6276.000000   3171.0
28   1.0  1.0  2870.0  10706.0  10340.000000   2731.0
29   0.0  1.0  3209.0   6090.0  13862.000000    436.0
..   ...  ...     ...      ...           ...      ...
970  1.0  1.0  2895.0  14756.0  11245.000000   4850.0
971  0.0  0.0  5620.0   1710.0   2594.000000   5948.0
972  1.0  0.0  4590.0   6185.0   7901.000000   5543.0
973  1.0  1.0  3560.0  10660.0  12264.000000   2016.0
974  0.0  0.0  1888.0   2047.0   2341.000000   2493.0
975  0.0  1.0  3786.0   5626.0  10650.000000    952.0
976  1.0  1.0  4511.0   9841.5  12187.000000   4015.0
977  1.0  1.0  2001.0  14057.0  10422.000000   4158.0
978  0.0  0.0  7890.0   1754.0   2150.000000   9262.0
979  0.0  1.0  4989.0   6121.0  13262.000000    232.0
980  0.0  1.0  1897.0   6345.0  13365.000000   1193.0
981  0.0  1.0  3881.0   6734.0  12585.000000   1359.0
982  0.0  0.0  8657.0   3242.0   4719.000000   9211.0
983  0.0  1.0  2805.0   6204.0  13575.000000    752.0
984  1.0  0.0  2608.0   4180.0   6532.000000   2476.0
985  0.0  0.0  7085.0   2156.0   3687.000000   6943.0
986  1.0  0.0  3797.0  13982.0   7208.000000   5746.0
987  0.0  1.0  1449.0   8922.0  11458.000000   4549.0
988  0.0  0.0  9087.0   1472.0   4954.000000   6449.0
989  0.0  0.0  4755.0   6829.0   8426.500000   1396.0
990  1.0  1.0  3935.0  12186.0  11899.000000   3602.0
991  1.0  1.0  3341.0  13339.0  13240.000000   2962.0
992  0.0  1.0  4656.0   8534.0  12789.000000   2894.0
993  0.0  0.0  4352.0   1792.0   2959.000000   4119.0
994  1.0  1.0  3048.0   7782.5   5375.000000   3771.0
995  1.0  0.0  2161.0  13773.0   7791.000000   3372.0
996  0.0  1.0  3779.0   7132.0  14182.000000   1224.0
997  0.0  0.0  5919.0   2703.0   3578.000000   6640.0
998  1.0  1.0  3825.0  14157.0  10174.000000   5970.0
999  1.0  1.0  3847.0  14364.0  14690.000000   3117.0

[1000 rows x 6 columns]

In [24]:
DF3 == DF4
DF2 == DF5

Q1    Q2     X1     X2     X3     X4
0    True  True   True   True   True   True
1    True  True   True   True   True   True
2    True  True   True   True   True   True
3    True  True   True  False   True   True
4    True  True   True   True   True   True
5    True  True   True   True   True   True
6    True  True   True   True   True   True
7    True  True   True   True   True   True
8    True  True   True   True   True   True
9    True  True   True  False   True   True
10   True  True   True   True   True   True
11   True  True   True   True   True   True
12   True  True   True   True   True   True
13   True  True   True   True   True   True
14   True  True   True   True   True   True
15   True  True   True   True   True   True
16   True  True   True   True   True   True
17   True  True   True   True   True   True
18   True  True   True   True   True   True
19   True  True   True   True   True   True
20   True  True   True   True   True   True
21   True  True   True   True   True   True
22   True  True   True   True   True   True
23   True  True   True   True   True   True
24   True  True   True   True   True   True
25   True  True   True  False  False   True
26   True  True   True   True  False   True
27   True  True   True  False   True   True
28   True  True   True   True   True   True
29   True  True   True   True   True   True
..    ...   ...    ...    ...    ...    ...
970  True  True   True   True   True   True
971  True  True   True   True   True   True
972  True  True  False  False   True   True
973  True  True   True   True   True   True
974  True  True   True   True   True   True
975  True  True   True   True   True   True
976  True  True   True  False   True   True
977  True  True   True   True   True   True
978  True  True   True   True   True   True
979  True  True   True   True   True   True
980  True  True   True   True   True   True
981  True  True   True   True   True   True
982  True  True   True   True   True   True
983  True  True   True   True   True   True
984  True  True   True  False   True   True
985  True  True   True   True   True   True
986  True  True   True   True   True  False
987  True  True   True   True   True   True
988  True  True   True   True   True   True
989  True  True   True  False  False   True
990  True  True   True   True   True   True
991  True  True   True   True   True   True
992  True  True   True   True   True   True
993  True  True   True   True   True   True
994  True  True   True  False  False   True
995  True  True   True   True   True   True
996  True  True   True   True   True   True
997  True  True   True   True   True   True
998  True  True   True   True   True   True
999  True  True   True   True   True   True

[1000 rows x 6 columns]

Df2: threshold check, mean
Df3: bfill
Df4: ffill
Df5: threshold check, median
Df6: threshold check, linear interpolation
